topdown/reachable.go

```diff
@@ -56,6 +69,70 @@ func builtinReachable(bctx BuiltinContext, args []*ast.Term, iter func(*ast.Term
 	return iter(ast.NewTerm(reached))
 }
 
+// pathBuilder is called recursively to build an array of paths that are reachable from the root
+func pathBuilder(graph ast.Object, root *ast.Term, path []*ast.Term, paths []*ast.Term, reached ast.Set) []*ast.Term {
+	if edges := graph.Get(root); edges != nil {
+		path = append(path, root)
+
+		if numberOfEdges(edges) >= 1 {
+			foreachVertex(edges, func(neighbor *ast.Term) {
+				if reached.Contains(neighbor) {
+					// If we've already reached this node, return current path (avoid infinite recursion)
+					paths = append(paths, path...)
+				} else {
+					reached.Add(root)
+					paths = pathBuilder(graph, neighbor, path, paths, reached)
+				}
+			})
+		} else {
+			paths = append(paths, path...)
+		}
+	} else {
+		// Node is invalid (not in graph). Commit the current path (without adding this root)
+		paths = append(paths, path...)
+	}
+
+	return paths
+}
+
+func builtinReachablePaths(bctx BuiltinContext, args []*ast.Term, iter func(*ast.Term) error) error {
+	// Return an error if the first argument is not an object.
+	graph, err := builtins.ObjectOperand(args[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// This is a queue that holds all nodes we still need to visit.  It is
+	// initialised to the initial set of nodes we start out with.
+	var queue []*ast.Term
+	foreachVertex(args[1], func(t *ast.Term) {
+		queue = append(queue, t)
+	})
+
+	results := ast.NewSet()
+
+	for len(queue) > 0 {
```

Comment:
Just for my understanding, it's the same as 
```
for _, node := range queue {
// ...
```
or is it not? We're not adding to `queue` in the loop body, are we?

Reply:
To be honest, this was just copied logic from `builtinReachable`, but you're right... using a range should be preferable here.